fastlane/lib/fastlane/actions/slack.rb

```diff
@@ -1,70 +1,178 @@
+require 'fastlane/notification/slack'
+
 # rubocop:disable Style/CaseEquality
 # rubocop:disable Style/MultilineTernaryOperator
 # rubocop:disable Style/NestedTernaryOperator
 module Fastlane
   module Actions
     class SlackAction < Action
-      def self.is_supported?(platform)
-        true
-      end
+      class Runner
+        def initialize(slack_url)
+          @notifier = Fastlane::Notification::Slack.new(slack_url)
+        end
 
-      # As there is a text limit in the notifications, we are
-      # usually interested in the last part of the message
-      # e.g. for tests
-      def self.trim_message(message)
-        # We want the last 7000 characters, instead of the first 7000, as the error is at the bottom
-        start_index = [message.length - 7000, 0].max
-        message = message[start_index..-1]
-        # We want line breaks to be shown on slack output so we replace
-        # input non-interpreted line break with interpreted line break
-        message.gsub('\n', "\n")
-      end
+        def run(options)
```

Comment:
First of all, Good to see, finally we are working on Slack notification improvement.

But I am afraid if this is the right approach or not. This is really breaking change. 
Now we have wrapped everything inside this Runner class and removed all helper methods. 😢

For example:
- **Earlier** SlackAction has helper method `def self.generate_slack_attachments(options)` and anyone can call it directly to generate beautiful slack attachements i.e https://github.com/crazymanish/fastlane-plugin-slack_bot/blob/master/lib/fastlane/plugin/slack_bot/actions/post_to_slack.rb#L23
- **Now** SlackAction removed above and introduced `def generate_slack_attachments(options)` and to call this method we have to init runner first `Runner.new(options[:slack_url]).generate_slack_attachments(options)`

This is a breaking change, why SlackAction need Runner with slack_url to generate slack attachements
We have combined two things into one:
- Notifier just required to post a slack message, that's all.
- Generate slack attachments should not be dependent on the notifier

I am requesting you, please think about it! 🙇‍♂️😇

Otherwise, the User has to do some hack to access the `def generate_slack_attachments(options)` method i.e
```ruby
def self.generate_slack_attachments(options)
   fake_url = "https://this.is.fake.url.to.generate.attachments.com"
   Runner.new(fake_url).generate_slack_attachments(options)
 end
```

<img width="524" alt="Screenshot 2021-04-10 at 12 47 18" src="https://user-images.githubusercontent.com/5364500/114267798-22d41780-99fe-11eb-990d-38ddcaec7eb9.png">

Reply:
Hey! Thank you for your feedback and sorry for making you confused but that breaking change was made in https://github.com/fastlane/fastlane/pull/18512 and this PR is subsequent PR of it. I couldn't open a PR targeting #18512's branch as both sent from my forked repo😅 

> This is really breaking change.
> Now we have wrapped everything inside this Runner class and removed all helper methods. 😢

Regarding the breaking you pointed, I can keep it working for now, in short. The purpose of the extraction is to make it testable without hacky way as well as making existing testing specs independent from `slack-notifier` gem. I didn't intend such disruption for any users🙂 

However, in _my personal_ opinion, fastlane focuses on providing DSL for mobile apps CI/CD rather than being a Ruby library for writing tools. Especially, classes sitting under `Fastlane::Actions` namespace, it's never meant to be libraries directly accessed from other gems, I guess. So I don't recommend you reuse methods implemented in actions. That said, as far as I'm aware there's no such policy explicitly declared and in fact it was not a private method so nothing prevents you from reusing them on your own. 

Another reason that I don't recommend is that as I explained in the PR description, Slack's attachment is now _almost_ deprecated https://api.slack.com/messaging/composing/layouts#when-to-use-attachments. While it will be working for a while it's good to migrate the attachment format to the new Block Kit format. I am thinking of doing the migration near future for fastlane too.

What do you think about keeping the helpers working with deprecation message and remove them when migrating to Block Kit? If you still want exactly the same behaviour in your gem, it's up to you to copy existing logic in fastlane to your gem.

Reply:
I'll update #18512 to keep `generate_slack_attachments` or other helpers working.

Reply:
@crazymanish Can you please have a look at this and leave any feedback there?🙂 

https://github.com/fastlane/fastlane/pull/18512/files#diff-e17259d099a0447ae1c18920e3251ec219f6069913276bf742f6687d9ea8ff80R299-R301

Reply:
> What do you think about keeping the helpers working with deprecation message and remove them when migrating to Block Kit?

Thank you very much @ainame I love this approach. 🙇 🙇‍♂️ ❤️ 
Now, instead of introducing breaking change directly, now first we are introducing deprecation message and later breaking changes, which is really cool. This deprecation message will give time to users to perform the migration. 
Thank you very much again! 👏🙌

> However, in my personal opinion, fastlane focuses on providing DSL for mobile apps CI/CD rather than being a Ruby library for writing tools. Especially, classes sitting under Fastlane::Actions namespace, it's never meant to be libraries directly accessed from other gems, I guess. 

Yes, I agree!
[fastlane-plugin-slack_bot](https://github.com/crazymanish/fastlane-plugin-slack_bot) gem is fastlane-plugin (because Fastlane don't accept lots of new actions anymore) so Idea was to reuse Fastlane logic as much as possible and keep things consistent without duplicate code 😊